server/src/main/java/org/opensearch/common/cache/stats/StatsHolder.java

```diff
@@ -0,0 +1,200 @@
+ *
+ * @opensearch.experimental
+ */
+public class StatsHolder {
+
+    // The list of permitted dimensions. Should be ordered from "outermost" to "innermost", as you would like to
+    // aggregate them in an API response.
+    private final List<String> dimensionNames;
+
+    // A tree structure based on dimension values, which stores stats values in its leaf nodes.
+    // Non-leaf nodes have stats matching the sum of their children.
+    private final DimensionNode statsRoot;
+
+    public StatsHolder(List<String> dimensionNames) {
+        this.dimensionNames = dimensionNames;
+        this.statsRoot = new DimensionNode(null); // The root node has no dimension value associated with it, only children
+        statsRoot.createChildrenMap();
+    }
+
+    public List<String> getDimensionNames() {
+        return dimensionNames;
+    }
+
+    // For all these increment functions, the dimensions list comes from the key, and contains all dimensions present in dimensionNames.
+    // The order has to match the order given in dimensionNames.
+    public void incrementHits(List<String> dimensionValues) {
+        internalIncrement(dimensionValues, (counter) -> counter.hits.inc(), true);
+    }
+
+    public void incrementMisses(List<String> dimensionValues) {
+        internalIncrement(dimensionValues, (counter) -> counter.misses.inc(), true);
+    }
+
+    public void incrementEvictions(List<String> dimensionValues) {
+        internalIncrement(dimensionValues, (counter) -> counter.evictions.inc(), true);
+    }
+
+    public void incrementSizeInBytes(List<String> dimensionValues, long amountBytes) {
+        internalIncrement(dimensionValues, (counter) -> counter.sizeInBytes.inc(amountBytes), true);
+    }
+
+    // For decrements, we should not create nodes if they are absent. This protects us from erroneously decrementing values for keys
+    // which have been entirely deleted, for example in an async removal listener.
+    public void decrementSizeInBytes(List<String> dimensionValues, long amountBytes) {
+        internalIncrement(dimensionValues, (counter) -> counter.sizeInBytes.dec(amountBytes), false);
+    }
+
+    public void incrementEntries(List<String> dimensionValues) {
+        internalIncrement(dimensionValues, (counter) -> counter.entries.inc(), true);
+    }
+
+    public void decrementEntries(List<String> dimensionValues) {
+        internalIncrement(dimensionValues, (counter) -> counter.entries.dec(), false);
+    }
+
+    /**
+     * Reset number of entries and memory size when all keys leave the cache, but don't reset hit/miss/eviction numbers.
+     * This is in line with the behavior of the existing API when caches are cleared.
+     */
+    public void reset() {
+        resetHelper(statsRoot);
+    }
+
+    private void resetHelper(DimensionNode current) {
+        CacheStatsCounter counter = current.getStats();
+        counter.sizeInBytes.dec(counter.getSizeInBytes());
+        counter.entries.dec(counter.getEntries());
```

Comment:
Stats should expose reset operation for these 2 values. The dec operation can potentially make values negative, hence better to set 0

Reply:
We can add this for reset, but we won't totally avoid decrementing, for example when a key exits the cache and we decrement entries/sizeInBytes.

Reply:
Yeah, agree with that. Need to cover that with other synchronization concerns